### PR TITLE
Fix for IDENTITY-3515

### PR DESCRIPTION
**Pull request by ChamaraPhilipsuom**
Provided the option to add multiple destination urls.

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationConstants.java [hunk 1]
```diff
@@ -116,6 +116,7 @@ private OpenID() {
             public static final String AUTHENTICATION_CONTEXT_CLASS = "AuthnContextClassRef";
             public static final String AUTHENTICATION_CONTEXT_COMPARISON_LEVEL = "AuthnContextComparisonLevel";
             public static final String ATTRIBUTE_CONSUMING_SERVICE_INDEX = "AttributeConsumingServiceIndex";
+            public static final String DESTINATION_URL_PREFIX = "DestinationURL";
```
Comment (johannnallathamby) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationConstants.java]: Must Destination be a URL ?

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/java/org/wso2/carbon/idp/mgt/ui/util/IdPManagementUIUtil.java [hunk 2]
```diff
@@ -1677,4 +1677,17 @@ private static void buildInboundProvisioningConfiguration(IdentityProvider fedId
         }
         return null;
     }
+
+    public static List<org.wso2.carbon.identity.application.common.model.idp.xsd.Property> getPropertySetStartsWith(
```
Comment (johannnallathamby) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/java/org/wso2/carbon/idp/mgt/ui/util/IdPManagementUIUtil.java]: Can move this method to IdentityApplicationManagementUtil.

File: components/sso-saml/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java [hunk 3]
```diff
@@ -490,6 +491,66 @@ public static Issuer getIssuerFromTenantDomain(String tenantDomain) throws Ident
...
+            if (IdentityApplicationConstants.Authenticator.SAML2SSO.NAME.equals(config.getName())) {
+                for (Property prop : config.getProperties()) {
+                    if (prop.getName().startsWith(IdentityApplicationConstants.Authenticator.SAML2SSO
+                            .DESTINATION_URL_PREFIX)) {
+                        destinationURLs.add(prop.getValue());
+                    }
+                }
+
+            }
+        }
```
Comment (johannnallathamby) [inline components/sso-saml/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java]: Move this looping to IdentityApplicationManagementUtil

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationConstants.java (hunk 1 above)
Comment (ChamaraPhilipsuom) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationConstants.java]: It should be a URI.

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationManagementUtil.java [hunk 4]
```diff
@@ -1153,4 +1162,45 @@ public static String getDigestAlgoURIByConfig() {
...
+    }
+
+    /**
+     * This is used in back end. Property is the type of stub generated property
+     * @param authnConfigs authenticatorConfigs to iterate
+     * @param authenticatorName authenticator name of which the values are needed
+     * @param propNameStartsWith the prefix of the property name
+     * @return the list of values which statrts with the propNameStartsWith.
+     */
+    public static List<String> getPropertyValuesForNameStartsWith(FederatedAuthenticatorConfig[] authnConfigs, String authenticatorName, String propNameStartsWith){
```
Comment (omindu) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationManagementUtil.java]: Formatting issue. Also check the line width.

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp [hunk 5]
```diff
@@ -52,6 +56,7 @@
     String stsUrl = null;
     String sessionIdleTimeout = null;
     String rememberMeTimeout = null;
+    List<Property> destinationURLList = new ArrayList<Property>();
     FederatedAuthenticatorConfig[] federatedAuthenticators = residentIdentityProvider.getFederatedAuthenticatorConfigs();
     for(FederatedAuthenticatorConfig federatedAuthenticator : federatedAuthenticators){
```
Comment (omindu) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp]: Formatting issue.

Comment (ChamaraPhilipsuom) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp]: Fixed at #1682

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationManagementUtil.java (hunk 4 above)
Comment (ChamaraPhilipsuom) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationManagementUtil.java]: Fixed at #1682

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local.jsp [hunk 6]
```diff
@@ -48,11 +48,21 @@
         identityProvider.setHomeRealmId(request.getParameter("homeRealmId"));
         FederatedAuthenticatorConfig samlFedAuthn = new FederatedAuthenticatorConfig();
         samlFedAuthn.setName(IdentityApplicationConstants.Authenticator.SAML2SSO.NAME);
-        Property[] properties = new Property[1];
+        String[] destinationUrls = request.getParameter("destinationURLs").split(",");
+        Property[] properties = new Property[1+destinationUrls.length];
```
Comment (omindu) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local.jsp]: Formatting issue.

Comment (ChamaraPhilipsuom) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local.jsp]: Fixed at #1682

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp [hunk 7]
```diff
@@ -65,6 +70,12 @@
                     IdentityApplicationConstants.Authenticator.SAML2SSO.SSO_URL).getValue();
             samlSLOUrl = IdPManagementUIUtil.getProperty(properties,
                     IdentityApplicationConstants.Authenticator.SAML2SSO.LOGOUT_REQ_URL).getValue();
+            destinationURLList = IdentityApplicationManagementUtil.getPropertySetStartsWith(properties, IdentityApplicationConstants
+                    .Authenticator.SAML2SSO.DESTINATION_URL_PREFIX);
+            if (destinationURLList.size() == 0){
```
Comment (omindu) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp]: Formatting issue.

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp [hunk 8]
```diff
@@ -65,6 +70,12 @@
...
+                destinationURLList.add(IdPManagementUIUtil.getProperty(properties,
+                    IdentityApplicationConstants.Authenticator.SAML2SSO.SSO_URL));
+            }
         } else if(IdentityApplicationConstants.OAuth10A.NAME.equals(federatedAuthenticator.getName())){
```
Comment (omindu) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp]: Formatting issue.

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationManagementUtil.java [hunk 9]
```diff
@@ -1153,4 +1162,45 @@ public static String getDigestAlgoURIByConfig() {
...
+
+    /**
+     * This is used in front end. Property is the type of stub generated property
+     * @param properties properties list to iterate
+     * @param startWith the peoperty list startswith the given name
+     * @return
+     */
+    public static List<org.wso2.carbon.identity.application.common.model.idp.xsd.Property> getPropertySetStartsWith(
+            org.wso2.carbon.identity.application.common.model.idp.xsd.Property[] properties,
+            String startWith) {
```
Comment (pulasthi7) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationManagementUtil.java]: better to rename the variable to something like 'queyString'.

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local.jsp [hunk 10]
```diff
@@ -48,11 +48,21 @@
...
         Property property = new Property();
         property.setName(IdentityApplicationConstants.Authenticator.SAML2SSO.IDP_ENTITY_ID);
         property.setValue(request.getParameter("idPEntityId"));
         properties[0] = property;
+        if (destinationUrls != null && destinationUrls.length > 0) {
+            for (int destinationCount = 1; destinationCount <= destinationUrls.length; destinationCount++) {
+                property = new Property();
+                property.setName(IdentityApplicationConstants.Authenticator.SAML2SSO.DESTINATION_URL_PREFIX + "_" + destinationCount);
+                property.setValue(destinationUrls[destinationCount-1]);
```
Comment (pulasthi7) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local.jsp]: format code

Comment (ChamaraPhilipsuom) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-finish-local.jsp]: Fixed at #1682

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp (hunk 7 above)
Comment (ChamaraPhilipsuom) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp]: Fixed at #1682

File: components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp (hunk 8 above)
Comment (ChamaraPhilipsuom) [inline components/idp-mgt/org.wso2.carbon.idp.mgt.ui/src/main/resources/web/idpmgt/idp-mgt-edit-local.jsp]: Fixed at 1682

File: components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationManagementUtil.java (hunk 9 above)
Comment (ChamaraPhilipsuom) [inline components/application-mgt/org.wso2.carbon.identity.application.common/src/main/java/org/wso2/carbon/identity/application/common/util/IdentityApplicationManagementUtil.java]: This method no longer used.

**Pull request merged**